src/message/fetchActions.js

```diff
@@ -286,8 +286,16 @@ export const uploadFile = (narrow: Narrow, uri: string, name: string) => async (
   getState: GetState,
 ) => {
   const auth = getAuth(getState());
-  const response = await api.uploadFile(auth, uri, name);
-  const messageToSend = `[${name}](${response.uri})`;
 
-  dispatch(addToOutbox(narrow, messageToSend));
+  // @TODO translate the string 'Uploading file'.
+  const timestamp = await dispatch(addToOutbox(narrow, '[Uploading file...]', true));
+  let response;
+  try {
+    response = await api.uploadFile(auth, uri, name);
+  } catch (err) {
+    dispatch(deleteOutboxMessage(timestamp));
+    return;
+  }
+  const messageToSend = `[${name}](${response.uri})`;
+  dispatch(addToOutbox(narrow, messageToSend, false, timestamp));
```

Comment:
It makes me kind of nervous to be constructing this whole outbox object (and all its properties) a second time here. What if it gets some different answers the second time? Which should prevail? I'd rather not have to worry about that.

(Also it means doing a variety of work twice, which is the sort of thing that can contribute to making things slow. In this case I think there's nothing expensive that gets done twice, but because it's spread over a decent bit of code that *looks* like it gets called once per message sent, that could change innocently in the future.)

Instead let's have a new action type like UPDATE_OUTBOX_MESSAGE, which just contains a `timestamp` property (really `id` would be better, but the existing code uses `timestamp` as an ID and we can follow that for now), `deferSending`, and `content` and `markdownContent`.

Then instead of this second call to `addToOutbox`, we can call a new thunk action creator, something like `dispatch(completeOutboxMessage(timestamp, messageToSend));`, and that one will dispatch one of the new action type to update Redux and then dispatch a `sendOutbox()`.